orion-server/src/main/java/com/pinterest/orion/core/actions/kafka/ClusterRecoveryAction.java

```diff
@@ -0,0 +1,223 @@
+                markFailed(String.format(
+                        "ClusterRecoveryAction failed to recover brokers %s in cluster %s. ",
+                        candidates,
+                        cluster.getClusterId()));
+            }
+        } catch (Exception e) {
+            markFailed(e);
+        }
+    }
+
+    protected void healBroker(String deadBrokerId) throws Exception {
+        // This will trigger an action that will attempt to replace the broker.
+        // If agent is still online but Kafka process is down, it will try to restart the broker first.
+        Action brokerRecoveryAction = newBrokerRecoveryAction();
+        brokerRecoveryAction.setAttribute(OrionConstants.NODE_ID, deadBrokerId, sensorSet);
+        brokerRecoveryAction.setOwner("ClusterRecoveryAction");
+
+        if (nonExistentBrokers.contains(deadBrokerId)) {
+            Node existingNode = cluster.getNodeMap().values().iterator().next();
+            String extractedName = deriveNonexistentHostname(
+                    existingNode.getCurrentNodeInfo().getHostname(),
+                    existingNode.getCurrentNodeInfo().getNodeId(),
+                    deadBrokerId
+            );
+            // setting these attributes to indicate that the node doesn't exist in cluster map, and should skip any node-related checks
+            brokerRecoveryAction.setAttribute(BrokerRecoveryAction.ATTR_NODE_EXISTS_KEY, false);
+            brokerRecoveryAction.setAttribute(BrokerRecoveryAction.ATTR_NONEXISTENT_HOST_KEY, extractedName);
+        }
+        if (maybeDeadBrokers.contains(deadBrokerId)) {
+            // Setting this flag in the action will restart the broker before replacing the broker
+            brokerRecoveryAction.setAttribute(BrokerRecoveryAction.ATTR_TRY_TO_RESTART_KEY, true);
+            String restartNote = "Will try to restart node " + deadBrokerId + " before replacing it. ";
+            logger.info(restartNote);
+            getResult().appendOut(restartNote);
+        }
+        String dispatchNote = String.format(
+                "Dispatching BrokerRecoveryAction for node %s in cluster %s. ",
+                deadBrokerId,
+                cluster.getClusterId());
+        logger.info(dispatchNote);
+        getResult().appendOut(dispatchNote);
+        getEngine().dispatch(brokerRecoveryAction);
+    }
+
+    protected boolean healBrokers(Set<String> candidates) throws Exception {
+        String output;
+        boolean isSucceed = false;
+        if (candidates.size() == 1) {
+            output = String.format(
+                    "ClusterRecoveryAction trys to recover broker %s. ",
+                    candidates.iterator().next());
+            String deadBrokerId = candidates.iterator().next();
+            healBroker(deadBrokerId);
+            logger.info(output);
+            isSucceed = true;
+        } else if (candidates.size() > 1){
+            // more than 1 brokers are dead... better alert and have human intervention
```

Comment:
Wasn't the goal to heal multiple dead brokers automatically too?

Reply:
The ClusterRecoveryAction will keep the original workflow. The new workflow will be updated in child class.